sunholmexp.py

```diff
@@ -378,6 +408,48 @@ def process_bonus_exp_event(event: Any, state: State) -> List[str]:
     ]
 
 
+################################################################################
+def process_levelup_event(event: Any, state: State) -> List[str]:
+    name = event["name"]
+    level_change = event["levels"]
+    current_exp = state.players[name]
+    current_level = get_level_from_exp(current_exp)
+
+    if name not in state.players:
+        print("WARNING: Player not found for bonus", event)
+        return []
+
+    # TODO: explicitly support level penalties.
+    if level_change < 1:
+        print("WARNING: Levelup levels less than 1", event)
+        return []
+
+    # TODO: should this level up as high as possible or error out?
+    if current_level + level_change >= MAX_LEVEL:
+        print("WARNING: Levelup would exceed max", event)
+        return []
+
+    current_level_min = level_exp_caps[current_level - 1] # Always at least lvl 1.
+    current_level_max = level_exp_caps[current_level]
+    intended_level_min = level_exp_caps[current_level + level_change - 1]
+    intended_level_max = level_exp_caps[current_level + level_change]
+
+    preserved_exp = 0
+    if event["preserve"]:
+        progress = (current_exp - current_level_min) / (current_level_max - current_level_min)
+        preserved_exp = progress * (intended_level_max - intended_level_min)
+
+    state.players[name] += intended_level_min + preserved_exp
```

Comment:
Unironically, I think this should be =, not += since you want to set the player's new experience exactly at (intended_level_min + preserved_exp)

Reply:
Yep, refactored and looked over this, ty